File: rust/src/benchmark/user_worker.rs
use crate::benchmark::task::{Task, TaskResult};
use crate::api::ExecutorApi;
use crate::model::domain::{Operation, Period, User};
use crate::model::metrics::{Action, DomainMetric};
use std::sync::Arc;
use std::time::SystemTime;
use async_channel::{Receiver, Sender};

pub async fn user_worker(
    user: User,
    tasks: Receiver<Task>,
    results: Sender<TaskResult>,
    executor: Arc<dyn ExecutorApi>,
) {
    while let Ok(task) = tasks.recv().await {
        let start = SystemTime::now();
        let result = process_operation(user, task.operation, executor.clone()).await;
        let duration = start.elapsed().unwrap().as_secs_f64();
        let metric = DomainMetric {
            year: result.year,
            period: Some(result.period),
            index: task.index,
            user_no: user.0,
            action: result.action,
            duration,
        };
        results
            .send(TaskResult(metric))
            .await
            .expect(format!("Error sending task result from user {}", user.id()).as_str());
    }
}

async fn process_operation(
    user: User,
    operation: Operation,
    executor: Arc<dyn ExecutorApi>,
) -> OperationResult {
    match operation {
        Operation::Purchase(operation) => {
            executor.purchase_material(&operation, user).await;
            let yp = Period::from(operation.posting_date);
            OperationResult::new(yp.year(), yp.month(), Action::Purchase(operation))
        }
        Operation::Sale(operation) => {
            executor.sell_material(&operation, user).await;
            let yp = Period::from(operation.posting_date);
            OperationResult::new(yp.year(), yp.month(), Action::Sale(operation))
        }
        Operation::Cost(operation) => {
            executor.account_cost(&operation, user).await;
            let yp = Period::from(operation.posting_date);
            OperationResult::new(yp.year(), yp.month(), Action::Cost(operation))
        }
        Operation::PeriodReport(period) => {
            executor.period_report(period).await;
            OperationResult::new(period.year(), period.month(), Action::PeriodReport(period))
        }
        Operation::YearReport(period) => {
            executor.year_report(period).await;
            OperationResult::new(period.year(), period.month(), Action::YearReport(period))
        }
        Operation::OpenPeriod(period) => {
            let start = SystemTime::now();
            executor.open_period(period, user).await;
            let duration = start.elapsed().unwrap();
            println!("Open period {} done in {:?}", period.next_period(), duration);
            OperationResult::new(period.year(), period.month(), Action::OpenPeriod(period))
        }
        Operation::ClosePeriod(period) => {
            let start = SystemTime::now();
            executor.close_period(period, user).await;
            let duration = start.elapsed().unwrap();
            println!("Closing period {} done in {:?}", period.prev_period(), duration);
            OperationResult::new(period.year(), period.month(), Action::ClosePeriod(period))
        }
    }
}

struct OperationResult {
    pub year: u32,
    pub period: u32,
    pub action: Action,
}

impl OperationResult {
    fn new(year: u32, period: u32, action: Action) -> Self {
        OperationResult {
            year,
            period,
            action,
        }
    }
}

File: golang/benchmark/user_worker.go
package benchmark

import (
	"bench-pg-go/api"
	"bench-pg-go/model/domain"
	"bench-pg-go/model/metrics"
	"fmt"
	"sync"
	"time"
)

type operationResult struct {
	year   int
	month  int
	action metrics.Action
}

func userWorker(user domain.User, executor api.ExecutorApi, tasks chan Task, results chan TaskResult, wg *sync.WaitGroup) {
	defer wg.Done()
	for task := range tasks {

		start := time.Now()
		result := processOperation(user, task.Operation, executor)
		duration := time.Now().Sub(start)

		results <- TaskResult{
			metrics.DomainMetric{
				Year:     result.year,
				Period:   &result.month,
				Index:    task.Index,
				UserNo:   user.No(),
				Action:   result.action,
				Duration: duration.Seconds(),
			},
		}
	}
}

func processOperation(
	user domain.User,
	operation domain.Operation,
	executor api.ExecutorApi,
) operationResult {
	var result operationResult
	switch op := operation.(type) {
	case domain.Purchase:
		executor.PurchaseMaterial(&op, user)
		p := domain.PeriodFromDate(op.PostingDate)
		result = operationResult{p.Year(), p.Month(), metrics.Operation{op}}
	case domain.Sale:
		executor.SellMaterial(&op, user)
		p := domain.PeriodFromDate(op.PostingDate)
		result = operationResult{p.Year(), p.Month(), metrics.Operation{op}}
	case domain.Cost:
		executor.AccountCost(&op, user)
		p := domain.PeriodFromDate(op.PostingDate)
		result = operationResult{p.Year(), p.Month(), metrics.Operation{op}}
	case domain.PeriodReport:
		executor.PeriodReport(op.Period)
		result = operationResult{op.Period.Year(), op.Period.Month(), metrics.Operation{op}}
	case domain.YearReport:
		executor.YearReport(op.Period)
		result = operationResult{op.Period.Year(), op.Period.Month(), metrics.Operation{op}}
	case domain.OpenPeriod:
		start := time.Now()
		executor.OpenPeriod(op.Period, user)
		duration := time.Now().Sub(start)
		fmt.Printf("Open period %s done in %f\n", op.Period.NextPeriod().String(), duration.Seconds())
		result = operationResult{op.Period.Year(), op.Period.Month(), metrics.Operation{op}}
	case domain.ClosePeriod:
		start := time.Now()
		executor.ClosePeriod(op.Period, user)
		duration := time.Now().Sub(start)
		fmt.Printf("Close period %s done in %f\n", op.Period.PrevPeriod().String(), duration.Seconds())
		result = operationResult{op.Period.Year(), op.Period.Month(), metrics.Operation{op}}
	}
	return result
}
